File: 2kyu/symbolic-differentiation-of-prefix-expressions.rs
// https://www.codewars.com/kata/symbolic-differentiation-of-prefix-expressions/train/rust
// I struggled on this kata for such a long time because of a small pitfall.

use std::ops;
use std::fmt::Display;
use std::str::FromStr;

#[derive(Copy, Clone)]
enum UnaryOp {
    Sine,
    Cosine,
    Tangent,
    Exponent,
    Logarithm,
}

impl UnaryOp {
    fn eval(&self, x: f64) -> f64 {
        match self {
            UnaryOp::Sine => x.sin(),
            UnaryOp::Cosine => x.cos(),
            UnaryOp::Tangent => x.tan(),
            UnaryOp::Exponent => x.exp(),
            UnaryOp::Logarithm => x.ln(),
        }
    }
}

impl Display for UnaryOp {
    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
        write!(f, "{}", match self {
            UnaryOp::Sine => "sin",
            UnaryOp::Cosine => "cos",
            UnaryOp::Tangent => "tan",
            UnaryOp::Exponent => "exp",
            UnaryOp::Logarithm => "ln",
        })
    }
}

#[derive(Copy, Clone)]
enum BinaryOp {
    Add,
    Sub,
    Mul,
    Div,
    Pow,
}

impl BinaryOp {
    fn eval(&self, x: f64, y: f64) -> f64 {
        match self {
            BinaryOp::Add => x + y,
            BinaryOp::Sub => x - y,
            BinaryOp::Mul => x * y,
            BinaryOp::Div => x / y,
            BinaryOp::Pow => x.powf(y),
        }
    }
}

impl Display for BinaryOp {
    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
        write!(f, "{}", match self {
            BinaryOp::Add => '+',
            BinaryOp::Sub => '-',
            BinaryOp::Mul => '*',
            BinaryOp::Div => '/',
            BinaryOp::Pow => '^',
        })
    }
}

#[derive(Clone)]
enum Expr {
    Number(f64),
    Variable(char),
    Unary(UnaryOp, Box<Expr>),
    Binary(BinaryOp, Box<Expr>, Box<Expr>),
}

impl ops::Add<Expr> for Expr {
    type Output = Expr;

    fn add(self, rhs: Expr) -> Self::Output {
        Expr::Binary(BinaryOp::Add, Box::new(self), Box::new(rhs))
    }
}

impl ops::Sub<Expr> for Expr {
    type Output = Expr;

    fn sub(self, rhs: Expr) -> Self::Output {
        Expr::Binary(BinaryOp::Sub, Box::new(self), Box::new(rhs))
    }
}

impl ops::Mul<Expr> for Expr {
    type Output = Expr;

    fn mul(self, rhs: Expr) -> Self::Output {
        Expr::Binary(BinaryOp::Mul, Box::new(self), Box::new(rhs))
    }
}

impl ops::Div<Expr> for Expr {
    type Output = Expr;

    fn div(self, rhs: Expr) -> Self::Output {
        Expr::Binary(BinaryOp::Div, Box::new(self), Box::new(rhs))
    }
}

impl ops::BitXor<Expr> for Expr {
    type Output = Expr;

    fn bitxor(self, rhs: Expr) -> Self::Output {
        Expr::Binary(BinaryOp::Pow, Box::new(self), Box::new(rhs))
    }
}

impl Expr {
    fn zero() -> Expr { Expr::Number(0f64) }

    fn one() -> Expr { Expr::Number(1f64) }

    fn var(c: char) -> Expr { Expr::Variable(c) }

    fn is_constant(&self) -> bool {
        match self {
            Expr::Number(_) => true,
            Expr::Variable(_) => false,
            Expr::Unary(_, arg) => arg.is_constant(),
            Expr::Binary(_, lhs, rhs) => lhs.is_constant() && rhs.is_constant(),
        }
    }

    fn get_copy(&self) -> Expr {
        self.clone()
    }

    fn unary_with(&self, op: UnaryOp) -> Expr {
        Expr::Unary(op, Box::new(self.clone()))
    }

    fn sin(&self) -> Expr {
        self.unary_with(UnaryOp::Sine)
    }

    fn cos(&self) -> Expr {
        self.unary_with(UnaryOp::Cosine)
    }

    fn log(&self) -> Expr {
        self.unary_with(UnaryOp::Logarithm)
    }

    fn exp(&self) -> Expr {
        self.unary_with(UnaryOp::Exponent)
    }

    fn inverse(&self) -> Expr {
        Expr::one() / self.clone()
    }

    fn square(&self) -> Expr {
        Expr::Binary(BinaryOp::Pow, Box::new(self.clone()), Box::new(Expr::Number(2f64)))
    }

    fn neg(&self) -> Expr { Expr::Number(-1f64) * self.clone() }

    fn differentiate(&self) -> Expr {
        match self {
            Expr::Number(_) => Expr::zero(),
            Expr::Variable(_) => Expr::one(),
            Expr::Unary(op, arg) => match op {
                UnaryOp::Sine => arg.differentiate() * arg.cos(),
                UnaryOp::Cosine => arg.differentiate() * arg.sin().neg(),
                UnaryOp::Tangent => arg.differentiate() * (Expr::one() + self.square()),
                UnaryOp::Exponent => arg.differentiate() * self.get_copy(),
                UnaryOp::Logarithm => arg.differentiate() * arg.inverse(),
            },
            Expr::Binary(op, lhs, rhs) => match op {
                BinaryOp::Add => lhs.differentiate() + rhs.differentiate(),
                BinaryOp::Sub => lhs.differentiate() - rhs.differentiate(),
                BinaryOp::Mul => lhs.differentiate() * rhs.get_copy() + lhs.get_copy() * rhs.differentiate(),
                BinaryOp::Div => if rhs.is_constant() { // Denominator is constant
                    lhs.differentiate() / rhs.get_copy()
                } else { // General rule for division
                    (lhs.differentiate() * rhs.get_copy() - lhs.get_copy() * rhs.differentiate()) / rhs.square()
                },
                BinaryOp::Pow => match (lhs.is_constant(), rhs.is_constant()) { // the precedence here is the key to pass, notice parts embraced in brackets
                    (true, true) => Expr::zero(), // {a ^ b}' = 0
                    (true, false) => rhs.differentiate() * ((lhs.get_copy() ^ rhs.get_copy()) * lhs.log()), // {a ^ f(x)}' = f'(x) * [a ^ f(x) * ln a]
                    (false, true) => lhs.differentiate() * (rhs.get_copy() * (lhs.get_copy() ^ (rhs.get_copy() - Expr::one()))), // {f(x) ^ a}' = f'(x) * [a * f(x) ^ (a - 1)]
                    (false, false) => (rhs.get_copy() * lhs.log()).exp().differentiate(), // {f(x) ^ g(x)}' = {e ^ (g(x) * ln(f(x)))}'
                },
            },
        }
    }

    fn simplify(&self) -> Expr {
        match self {
            Expr::Unary(op, arg) => Expr::Unary(*op, Box::new(arg.simplify())),
            Expr::Binary(op, lhs, rhs) => match (op, lhs.simplify(), rhs.simplify()) {
                (op, Expr::Number(x), Expr::Number(y)) => Expr::Number(op.eval(x, y)),
                (BinaryOp::Add, Expr::Number(c), fx) => if c == 0f64 { fx } else { Expr::Number(c) + fx },
                (BinaryOp::Add, fx, Expr::Number(c)) => if c == 0f64 { fx } else { fx + Expr::Number(c) },
                (BinaryOp::Sub, Expr::Variable(a), Expr::Variable(b)) => if a == b { Expr::zero() } else { Expr::var(a) - Expr::var(b) }
                (BinaryOp::Sub, fx, Expr::Number(c)) => if c == 0f64 { fx } else { fx - Expr::Number(c) },
                (BinaryOp::Mul, fx, Expr::Number(c)) => if c == 0f64 { Expr::zero() } else if c == 1f64 { fx } else { fx * Expr::Number(c) },
                (BinaryOp::Mul, Expr::Number(c), fx) => if c == 0f64 { Expr::zero() } else if c == 1f64 { fx } else { Expr::Number(c) * fx }
                (BinaryOp::Div, Expr::Number(num), den) => if num == 0f64 { Expr::zero() } else { Expr::Number(num) / den },
                (BinaryOp::Div, num, Expr::Number(den)) => if den == 0f64 { panic!("divide by zero") } else if den == 1f64 { num } else { num / Expr::Number(den) },
                (BinaryOp::Pow, base, Expr::Number(exp)) => if exp == 0f64 { Expr::one() } else if exp == 1f64 { base } else { base ^ Expr::Number(exp) },
                (BinaryOp::Pow, Expr::Number(base), exp) => if base == 0f64 { Expr::zero() } else if base == 1f64 { Expr::one() } else { Expr::Number(base) ^ exp },
                (op, lhs, rhs) => Expr::Binary(*op, Box::new(lhs), Box::new(rhs)),
            },
            _ => self.clone(),
        }
    }
}

impl ToString for Expr {
    fn to_string(&self) -> String {
        match self {
            Expr::Number(value) => value.to_string(),
            Expr::Variable(name) => name.to_string(),
            Expr::Unary(op, arg) => format!("({} {})", op, arg.to_string()),
            Expr::Binary(op, lhs, rhs) => format!("({} {} {})", op, lhs.to_string(), rhs.to_string()),
        }
    }
}

impl FromStr for Expr {
    type Err = ExprParserError;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        let mut parser = ExprParser::new(s);
        parser.parse_expression()
    }
}

struct ExprParserError {
    at: usize,
    meet: Option<char>,
    message: &'static str,
}

impl ExprParserError {
    fn new(at: usize, meet: Option<char>, message: &'static str) -> ExprParserError {
        ExprParserError { at, meet, message }
    }
}

struct ExprParser {
    chars: Vec<char>,
    at: usize,
}

impl ExprParser {
    fn new(s: &str) -> ExprParser {
        ExprParser { chars: s.chars().collect(), at: 0 }
    }

    fn error(&self, message: &'static str) -> ExprParserError {
        let meet = if self.at < self.chars.len() { Some(self.chars[self.at]) } else { None };
        ExprParserError::new(self.at, meet, message)
    }

    fn peek(&self) -> Option<char> {
        if self.at < self.chars.len() {
            Some(self.chars[self.at])
        } else {
            None
        }
    }

    fn next(&mut self) -> Option<char> {
        if self.at < self.chars.len() {
            let save = self.chars[self.at];
            self.at += 1;
            Some(save)
        } else {
            None
        }
    }

    fn skip_whitespace(&mut self) {
        while let Some(c) = self.peek() {
            if c.is_whitespace() { self.next(); } else { break; }
        }
    }

    fn parse_expression(&mut self) -> Result<Expr, ExprParserError> {
        self.skip_whitespace();
        if let Some(c) = self.peek() {
            if c == '(' {
                self.next();
                self.skip_whitespace();
                let result = if let Some(op) = self.parse_unary() {
                    match self.parse_expression() {
                        Ok(arg) => Ok(Expr::Unary(op, Box::new(arg))),
                        e => e,
                    }
                } else if let Some(op) = self.parse_operator() {
                    self.next();
                    match self.parse_expression() {
                        Ok(lhs) => match self.parse_expression() {
                            Ok(rhs) => Ok(Expr::Binary(op, Box::new(lhs), Box::new(rhs))),
                            e => e,
                        },
                        e => e,
                    }
                } else {
                    Err(self.error("expect binary or unary expression"))
                };
                self.skip_whitespace();
                if let Some(')') = self.next() {
                    result
                } else {
                    Err(self.error("bracket doesn't match"))
                }
            } else if c.is_ascii_digit() || c == '.' || c == '-' {
                let mut buf = String::new();
                while let Some(c) = self.peek() {
                    if c.is_ascii_digit() || c == '.' || c == '-' {
                        buf.push(c);
                        self.next();
                    } else {
                        break;
                    }
                }
                Ok(Expr::Number(buf.parse::<f64>().unwrap()))
            } else if c.is_alphabetic() {
                self.next();
                Ok(Expr::Variable(c))
            } else {
                Err(self.error("unrecognized character"))
            }
        } else {
            Err(self.error("expect an expression instead of EOF"))
        }
    }

    fn parse_operator(&mut self) -> Option<BinaryOp> {
        match self.peek() {
            Some('+') => Some(BinaryOp::Add),
            Some('-') => Some(BinaryOp::Sub),
            Some('*') => Some(BinaryOp::Mul),
            Some('/') => Some(BinaryOp::Div),
            Some('^') => Some(BinaryOp::Pow),
            _ => None,
        }
    }

    fn parse_identifier(&mut self) -> Option<String> {
        let mut buf = String::new();
        while let Some(c) = self.peek() {
            if c.is_alphabetic() {
                buf.push(c);
                self.next();
            } else {
                break;
            }
        }
        if buf.is_empty() { None } else { Some(buf) }
    }

    fn parse_unary(&mut self) -> Option<UnaryOp> {
        if let Some(name) = self.parse_identifier() {
            match name.as_str() {
                "sin" => Some(UnaryOp::Sine),
                "cos" => Some(UnaryOp::Cosine),
                "tan" => Some(UnaryOp::Tangent),
                "exp" => Some(UnaryOp::Exponent),
                "ln" => Some(UnaryOp::Logarithm),
                _ => None,
            }
        } else {
            None
        }
    }
}

fn diff(expr: &str) -> String {
    match Expr::from_str(expr) {
        Ok(expr) => expr.simplify().differentiate().simplify().to_string(),
        Err(err) => if let Some(c) = err.meet {
            format!("{} of {}: {}", err.at, c, err.message)
        } else {
            format!("{}: {}", err.at, err.message)
        },
    }
}

File: 2kyu/symbolic-differentiation-of-prefix-expressions.js
class Expression {
  constructor() {}

  differentiate() {
    throw new Error("try to call `differentiate` on base class");
  }

  simplify() {
    throw new Error("try to call `simplify` on base class");
  }

  toString() {
    throw new Error("try to call `toString` on base class");
  }

  static parse(expr) {
    const NUMBER_PATTERN = /^([-+]?[0-9]*\.?[0-9]+(?:[eE][-+]?[0-9]+)?)/;
    const FUNCTION_PATTERN = /^\w+/;
    const OPERATOR_PATTERN = /^[\+\-\*\/\^]/;
    const WHITESPACE_PATTERN = /^\s+/;

    const FUNCION_MAP = {
      cos: Cosine,
      sin: Sine,
      tan: Tangent,
      ln: Logarithm,
      exp: Exponent
    };

    const OPERATOR_MAP = {
      "+": Add,
      "-": Subtract,
      "*": Multiply,
      "/": Divide,
      "^": Pow
    };

    function whitespace() {
      const result = expr.match(WHITESPACE_PATTERN);
      if (result) {
        expr = expr.substr(result[0].length);
      }
    }

    function expression() {
      let result = expr.match(NUMBER_PATTERN);
      if (result) {
        expr = expr.substr(result[0].length);
        return new Constant(parseFloat(result[0]));
      }
      if (expr.charAt(0) === "x") {
        expr = expr.substr(1);
        return new Variable();
      }
      if (expr.charAt(0) === "(") {
        let node;
        expr = expr.substr(1);
        whitespace();
        if ((result = expr.match(FUNCTION_PATTERN))) {
          expr = expr.substr(result[0].length);
          whitespace();
          const argument = expression();
          if (FUNCION_MAP.hasOwnProperty(result[0])) {
            node = new FUNCION_MAP[result[0]](argument);
          } else {
            throw new Error(`illegal function name "${result[0]}"`);
          }
        } else if ((result = expr.match(OPERATOR_PATTERN))) {
          expr = expr.substr(1);
          whitespace();
          const lhs = expression();
          whitespace();
          const rhs = expression();
          node = new OPERATOR_MAP[result[0]](lhs, rhs);
        } else {
          throw new Error(
            `expect an operator or a function name instead of "${expr}"`
          );
        }
        whitespace();
        if (expr.charAt(0) === ")") {
          expr = expr.substr(1);
        } else {
          throw new Error(`expect a ")" instead of "${expr}"`);
        }
        return node;
      }
      throw new Error(`expect a prefix expression instead of "${expr}"`);
    }

    const result = expression();
    whitespace();
    if (expr.length > 0) {
      throw new Error(`unrecognized character sequence ${expr}`);
    }
    return result;
  }
}

class Constant extends Expression {
  constructor(value) {
    super();
    this.value = value;
    Object.freeze(this);
  }

  differentiate() {
    return Constant.ZERO;
  }

  simplify() {
    return this;
  }

  toString() {
    return this.value.toString();
  }
}

Constant.ZERO = new Constant(0);
Constant.ONE = new Constant(1);

class Variable extends Expression {
  constructor() {
    super();
  }

  differentiate() {
    return Constant.ONE;
  }

  simplify() {
    return this;
  }

  toString() {
    return "x";
  }
}

class Add extends Expression {
  constructor(lhs, rhs) {
    super();
    this.lhs = lhs;
    this.rhs = rhs;
  }

  differentiate() {
    return new Add(this.lhs.differentiate(), this.rhs.differentiate());
  }

  simplify() {
    const lhs = this.lhs.simplify();
    const rhs = this.rhs.simplify();
    // constant folding
    if (lhs instanceof Constant) {
      if (lhs.value === 0) {
        return rhs;
      }
      if (rhs instanceof Constant) {
        return new Constant(lhs.value + rhs.value);
      }
    }
    if (rhs instanceof Constant) {
      if (rhs.value === 0) {
        return lhs;
      }
    }
    // no change
    if (lhs === this.lhs && rhs === this.rhs) {
      return this;
    }
    return new Add(lhs, rhs);
  }

  toString() {
    return `(+ ${this.lhs.toString()} ${this.rhs.toString()})`;
  }
}

class Subtract extends Expression {
  constructor(lhs, rhs) {
    super();
    this.lhs = lhs;
    this.rhs = rhs;
  }

  differentiate() {
    return new Subtract(this.lhs.differentiate(), this.rhs.differentiate());
  }

  simplify() {
    const lhs = this.lhs.simplify();
    const rhs = this.rhs.simplify();
    // x - x === 0
    if (lhs instanceof Variable && rhs instanceof Variable) {
      return new Constant(0);
    }
    // constant folding
    if (lhs instanceof Constant && rhs instanceof Constant) {
      if (rhs.value === 0) {
        return lhs;
      }
      return new Constant(lhs.value - rhs.value);
    }
    // no change
    if (lhs === this.lhs && rhs === this.rhs) {
      return this;
    }
    return new Subtract(lhs, rhs);
  }

  toString() {
    return `(- ${this.lhs.toString()} ${this.rhs.toString()})`;
  }
}

class Multiply extends Expression {
  constructor(lhs, rhs) {
    super();
    if (rhs instanceof Constant) {
      this.lhs = rhs;
      this.rhs = lhs;
    } else {
      this.lhs = lhs;
      this.rhs = rhs;
    }
  }

  differentiate() {
    return new Add(
      new Multiply(this.lhs.differentiate(), this.rhs),
      new Multiply(this.lhs, this.rhs.differentiate())
    );
  }

  simplify() {
    let lhs = this.lhs.simplify();
    let rhs = this.rhs.simplify();
    // put constant in the front
    if (rhs instanceof Constant) {
      const temp = lhs;
      lhs = rhs;
      rhs = temp;
    }
    // constant folding
    if (lhs instanceof Constant) {
      if (lhs.value === 0) {
        return Constant.ZERO;
      }
      if (lhs.value === 1) {
        return rhs;
      }
      if (rhs instanceof Constant) {
        return new Constant(lhs.value * rhs.value);
      }
    }
    // no change
    if (lhs === this.lhs && rhs === this.rhs) {
      return this;
    }
    return new Multiply(lhs, rhs);
  }

  toString() {
    return `(* ${this.lhs.toString()} ${this.rhs.toString()})`;
  }
}

class Divide extends Expression {
  constructor(lhs, rhs) {
    super();
    this.lhs = lhs;
    this.rhs = rhs;
  }

  differentiate() {
    // f(x) / a
    if (this.rhs instanceof Constant) {
      return new Divide(this.lhs.differentiate(), this.rhs);
    }
    // f(x) / g(x)
    return new Divide(
      new Subtract(
        new Multiply(this.lhs.differentiate(), this.rhs),
        new Multiply(this.lhs, this.rhs.differentiate())
      ),
      new Pow(this.rhs, new Constant(2))
    );
  }

  simplify() {
    const lhs = this.lhs.simplify();
    const rhs = this.rhs.simplify();
    if (lhs instanceof Constant) {
      if (lhs.value === 0) {
        return Constant.ZERO;
      }
      if (rhs instanceof Constant) {
        return new Constant(lhs.value / rhs.value);
      }
    }
    if (rhs instanceof Constant) {
      if (rhs.value === 0) {
        throw new Error("try to divide by zero");
      }
      if (rhs.value === 1) {
        return lhs;
      }
    }
    if (lhs === this.lhs && rhs === this.rhs) {
      return this;
    }
    return new Divide(lhs, rhs);
  }

  toString() {
    return `(/ ${this.lhs.toString()} ${this.rhs.toString()})`;
  }
}

class Pow extends Expression {
  constructor(base, exponent) {
    super();
    this.base = base;
    this.exponent = exponent;
  }

  differentiate() {
    // TODO use `hasVariable` instead
    // f(x) ^ a
    if (this.exponent instanceof Constant) {
      const a = this.exponent.value;
      if (a === 0) {
        return Constant.ZERO;
      }
      return new Multiply(
        new Constant(a),
        new Multiply(
          this.base.differentiate(),
          new Pow(this.base, new Constant(a - 1))
        )
      );
    }
    // a ^ f(x)
    if (this.base instanceof Constant) {
      const a = this.base.value;
      return new Multiply(
        new Logarithm(a),
        new Multiply(this.exponent.differentiate(), this)
      );
    }
    // f(x) ^ g(x)
    return new Exponent(
      new Multiply(this.exponent, new Logarithm(this.base))
    ).differentiate();
  }

  simplify() {
    const base = this.base.simplify();
    const exponent = this.exponent.simplify();
    if (base instanceof Constant) {
      if (exponent instanceof Constant) {
        return new Constant(Math.pow(base.value, exponent.value));
      }
      if (base.value === 1) {
        return Constant.ONE;
      }
    }
    if (exponent instanceof Constant) {
      // pow(x, 0) === 1
      if (exponent.value === 0) {
        return Constant.ONE;
      }
      // pow(x, 1) === x
      if (exponent.value === 1) {
        return base;
      }
    }
    // simplification doesn't change base and exponent
    if (base === this.base && exponent === this.exponent) {
      return this;
    }
    return new Pow(base, exponent);
  }

  toString() {
    return `(^ ${this.base.toString()} ${this.exponent.toString()})`;
  }
}

class Cosine extends Expression {
  constructor(argument) {
    super();
    this.argument = argument;
  }

  differentiate() {
    return new Multiply(
      this.argument.differentiate(),
      new Multiply(new Constant(-1), new Sine(this.argument))
    );
  }

  simplify() {
    const argument = this.argument.simplify();
    if (argument instanceof Constant) {
      return new Constant(Math.cos(argument.value));
    }
    if (argument === this.argument) {
      return this;
    }
    return new Cosine(argument);
  }

  toString() {
    return `(cos ${this.argument.toString()})`;
  }
}

class Sine extends Expression {
  constructor(argument) {
    super();
    this.argument = argument;
  }

  differentiate() {
    return new Multiply(
      this.argument.differentiate(),
      new Cosine(this.argument)
    );
  }

  simplify() {
    const argument = this.argument.simplify();
    if (argument instanceof Constant) {
      return new Constant(Math.sin(argument.value));
    }
    if (argument === this.argument) {
      return this;
    }
    return new Sine(argument);
  }

  toString() {
    return `(sin ${this.argument.toString()})`;
  }
}

class Tangent extends Expression {
  constructor(argument) {
    super();
    this.argument = argument;
  }

  differentiate() {
    return new Multiply(
      this.argument.differentiate(),
      new Add(new Constant(1), new Pow(this, new Constant(2)))
    );
  }

  simplify() {
    const argument = this.argument.simplify();
    if (argument instanceof Constant) {
      return new Constant(Math.tan(argument.value));
    }
    if (argument === this.argument) {
      return this;
    }
    return new Tangent(argument);
  }

  toString() {
    return `(tan ${this.argument.toString()})`;
  }
}

class Exponent extends Expression {
  constructor(argument) {
    super();
    this.argument = argument;
  }

  differentiate() {
    return new Multiply(this.argument.differentiate(), this);
  }

  simplify() {
    const argument = this.argument.simplify();
    if (argument instanceof Constant) {
      return new Constant(Math.exp(argument.value));
    }
    if (argument === this.argument) {
      return this;
    }
    return new Exponent(argument);
  }

  toString() {
    return `(exp ${this.argument.toString()})`;
  }
}

class Logarithm extends Expression {
  constructor(argument) {
    super();
    this.argument = argument;
  }

  differentiate() {
    return new Multiply(
      this.argument.differentiate(),
      new Divide(new Constant(1), this.argument)
    );
  }

  simplify() {
    const argument = this.argument.simplify();
    if (argument instanceof Constant) {
      return new Constant(Math.ln(argument.value));
    }
    if (argument === this.argument) {
      return this;
    }
    return new Logarithm(argument);
  }

  toString() {
    return `(exp ${this.argument.toString()})`;
  }
}

function checkConstructionAndStringify(expr) {
  const root = Expression.parse(expr);
  const serialized = root.toString();
  if (expr === serialized) {
    console.log(`Everything works well with "${expr}".`);
  } else {
    console.log(`Oops! "${expr}" !== "${serialized}".`);
  }
}

function diff(expr) {
  const y = Expression.parse(expr);
  console.log("The original tree:", y.toString());
  const dy = y.differentiate();
  console.log("Differentiated expression:", dy.toString());
  const reduced_dy = dy.simplify();
  console.log("Reduced differentiated expression:", reduced_dy.toString());
  return reduced_dy.toString();
}
